zkstack_cli/crates/zkstack/src/commands/dev/commands/v28_precompiles.rs

```diff
@@ -0,0 +1,293 @@
+        Self {
+            chain_id: value.chain_id,
+            gw_chain_id: value.gw_chain_id,
+            l1_rpc_url: value.l1_rpc_url,
+            gw_rpc_url: value.gw_rpc_url,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct V28UpgradeInfo {
+    // Information about pre-upgrade contracts.
+    l1_chain_id: u32,
+    gw_chain_id: u32,
+    pub(crate) bridgehub_addr: Address,
+    pub(crate) gw_bridgehub_addr: Address,
+
+    // Information from upgrade
+    chain_upgrade_diamond_cut: Bytes,
+
+    new_protocol_version: u64,
+    old_protocol_version: u64,
+
+    l1_gas_price: u64,
+
+    gateway_diamond_cut: Vec<u8>,
+}
+
+impl ZkStackConfig for V28UpgradeInfo {}
+
+pub(crate) async fn run(shell: &Shell, args: V28PrecompilesCalldataArgs) -> anyhow::Result<()> {
+    let forge_args = &Default::default();
+    let foundry_contracts_path = get_default_foundry_path(shell)?;
+
+    // 0. Read the GatewayUpgradeInfo
+
+    let upgrade_info = V28UpgradeInfo::read(shell, &args.upgrade_description_path)?;
+
+    // 1. Update all the configs
+
+    let chain_info = fetch_chain_info(&upgrade_info, &args.clone().into()).await?;
+
+    // 2. Generate calldata
+    let schedule_calldata = set_upgrade_timestamp_calldata(
+        upgrade_info.new_protocol_version,
+        args.server_upgrade_timestamp,
+    );
+
+    let set_timestamp_call = AdminCall {
+        description: "Calldata to schedule upgrade".to_string(),
+        data: schedule_calldata,
+        target: chain_info.chain_admin_addr,
+        value: U256::zero(),
+    };
+    println!("{}", serde_json::to_string_pretty(&set_timestamp_call)?);
+    println!("---------------------------");
+
+    if !args.force_display_finalization_params.unwrap_or_default() {
+        let chain_readiness = check_chain_readiness(
+            args.l1_rpc_url.clone(),
+            args.l2_rpc_url.clone(),
+            args.gw_rpc_url.clone(),
+            args.chain_id,
+            args.gw_chain_id,
+        )
+        .await;
+
+        if let Err(err) = chain_readiness {
+            print_error(err);
+            return Ok(());
+        };
+    }
+
+    let mut admin_calls_finalize = AdminCallBuilder::new(vec![]);
+
+    admin_calls_finalize.append_execute_upgrade(
+        chain_info.hyperchain_addr,
+        upgrade_info.old_protocol_version,
+        upgrade_info.chain_upgrade_diamond_cut.clone(),
+    );
+
+    admin_calls_finalize.display();
+
+    let (chain_admin_calldata, _) = admin_calls_finalize.compile_full_calldata();
+
+    println!(
+        "Full calldata to call `ChainAdmin` with : {}",
+        hex::encode(&chain_admin_calldata)
+    );
+
+    let mut admin_calls_gw = AdminCallBuilder::new(vec![]);
+
+    admin_calls_gw.append_execute_upgrade(
+        chain_info.gw_hyperchain_addr,
+        upgrade_info.old_protocol_version,
+        upgrade_info.chain_upgrade_diamond_cut.clone(),
+    );
+
+    admin_calls_gw.prepare_upgrade_chain_on_gateway_calls(
```

Comment:
why do you need both `append_execute_upgrade` and `prepare_upgrade_chain_on_gateway_calls`